File: js/test/helpers/factory.ts
import { dateTimeFromISO } from "../../dateTime";
import { Certification, CertificationData } from "../../models/certification";
import { Employee } from "../../models/employee";
import { OCSTrip } from "../../models/ocs";
import { StopTimeUpdate, TripUpdate } from "../../models/tripUpdate";
import { Vehicle } from "../../models/vehicle";
import { StopStatus, VehiclePosition } from "../../models/vehiclePosition";
import { Factory } from "fishery";
import { DateTime } from "luxon";

export const employeeFactory = Factory.define<Employee>(({ sequence }) => ({
  first_name: "Preferredy",
  last_name: "Lasty",
  badge: sequence.toString(),
}));

export const certificationDataFactory = Factory.define<CertificationData>(
  () => ({
    type: "rail",
    expires: "2023-12-12",
    rail_line: "blue",
  }),
);
export const certificationFactory = Factory.define<Certification>(() => ({
  type: "rail",
  expires: DateTime.fromISO("2023-12-12", { zone: "America/New_York" }),
  railLine: "blue",
}));

export const vehiclePositionFactory = Factory.define<VehiclePosition>(() => ({
  cars: ["1877", "1876", "1807", "1806", "1815", "1814"],
  directionId: 1,
  heading: 330,
  label: "1877",
  position: { latitude: 42.2773, longitude: -71.03159 },
  routeId: "Red",
  stationId: "place-jfk",
  stopId: "70096",
  stopStatus: StopStatus.InTransitTo,
  timestamp: dateTimeFromISO("2025-04-29T21:27:26.679Z"),
  vehicleId: "R-5482CAAA",
  tripId: "68077971",
}));

export const stopTimeUpdateFactory = Factory.define<StopTimeUpdate>(() => ({
  predictedArrivalTime: dateTimeFromISO("2025-04-29T21:51:38Z"),
  predictedDepartureTime: dateTimeFromISO("2025-04-29T21:53:38Z"),
  stationId: "place-brdwy",
}));

export const tripUpdateFactory = Factory.define<TripUpdate>(() => ({
  direction: 0,
  label: "1877",
  routeId: "Red",
  routePatternId: "Red-1-0",
  timestamp: dateTimeFromISO("2025-04-29T21:27:26.679Z"),
  tripId: "68078228",
  vehicleId: "R-54831F04",
  stopTimeUpdates: [stopTimeUpdateFactory.build()],
}));

export const ocsTripFactory = Factory.define<OCSTrip>(() => ({
  scheduledDeparture: dateTimeFromISO("2025-04-29T21:41:00.000Z"),
  scheduledArrival: dateTimeFromISO("2025-04-29T22:24:00.000Z"),
  actualDeparture: dateTimeFromISO("2025-04-29T21:43:00.000Z"),
  originStation: "ASHMONT",
  destinationStation: "ALEWIFE",
  offset: 0,
  deleted: false,
  updatedAt: dateTimeFromISO("2025-04-29T06:00:00.000Z"),
}));

export const vehicleFactory = Factory.define<Vehicle>(() => ({
  vehiclePosition: vehiclePositionFactory.build(),
  tripUpdate: tripUpdateFactory.build(),
  ocsTrips: {
    current: ocsTripFactory.build(),
    next: [],
  },
}));

File: test/support/factory.ex
defmodule Orbit.Factory do
  use ExMachina.Ecto, repo: Orbit.Repo

  def user_factory do
    %Orbit.Authentication.User{
      email: sequence(:user_email, &"fake#{&1}@test.com"),
      permissions: [],
      groups: []
    }
  end

  def employee_factory do
    %Orbit.Employee{
      first_name: "Fake",
      preferred_first: "Preferredy",
      last_name: "Person",
      middle_initial: "A",
      email: sequence(:employee_email, &"fake#{&1}@test.com"),
      badge_number: sequence("employee_badge"),
      badge_serials: [],
      area: 321
    }
  end

  def operator_sign_in_factory do
    {:ok, base_dt, 0} = DateTime.from_iso8601("2024-07-02T20:30:00Z")

    %Orbit.OperatorSignIn{
      signed_in_employee: build(:employee),
      signed_in_by_user: build(:user),
      signed_in_at: DateTime.add(base_dt, sequence(:operator_sign_in_time_offset, & &1), :minute),
      rail_line: :blue,
      radio_number: "22",
      sign_in_method: :manual
    }
  end

  def vehicle_position_factory do
    %Realtime.Data.VehiclePosition{
      route_id: :Red,
      revenue: true,
      direction: 0,
      label: "1700",
      position: %Util.Position{latitude: 42.3611328, longitude: -71.0706147},
      station_id: "place-chmnl",
      stop_id: "70073",
      current_status: :STOPPED_AT,
      timestamp: DateTime.from_unix!(1_640_000_000),
      vehicle_id: "R-547210A7"
    }
  end

  def trip_update_factory do
    %Realtime.Data.TripUpdate{
      label: "1700",
      route_id: :Red,
      direction: 0,
      trip_id: "68078228",
      route_pattern_id: "Red-1-0",
      vehicle_id: "R-547210A7",
      timestamp: DateTime.from_unix!(1_640_000_000),
      stop_time_updates: [build(:stop_time_update)]
    }
  end

  def stop_time_update_factory do
    %Realtime.Data.TripUpdate.StopTimeUpdate{
      station_id: "place-chmnl",
      predicted_arrival_time: DateTime.from_unix!(1_640_000_000),
      predicted_departure_time: DateTime.from_unix!(1_640_000_100)
    }
  end

  def tsch_new_factory do
    %Orbit.Ocs.Message.TschNewMessage{
      counter: 123,
      timestamp: DateTime.from_iso8601("2025-07-07T13:00:00.000-04:00") |> elem(1),
      transitline: "R",
      trip_uid: "11111",
      add_type: "S",
      trip_type: "G",
      ocs_route_id: "route-id",
      origin_sta: "origin-station",
      dest_sta: "destination-station",
      prev_trip_uid: "00000",
      next_trip_uid: "22222"
    }
  end

  def tsch_tag_factory do
    alias Orbit.Ocs.Message.TschTagMessage.CarTag

    %Orbit.Ocs.Message.TschTagMessage{
      counter: 123,
      timestamp: DateTime.from_iso8601("2025-07-07T13:00:00.000-04:00") |> elem(1),
      transitline: "R",
      trip_uid: "11111",
      train_uid: "22222",
      consist_tags: ["K", "N"],
      car_tags: [
        %CarTag{
          car_number: "1234",
          tag: "K"
        },
        %CarTag{
          car_number: "1235",
          tag: "K"
        }
      ]
    }
  end

  def trip_factory do
    %Orbit.Ocs.Trip{
      service_date: Date.from_iso8601!("2025-07-07"),
      uid: "11111",
      train_uid: "99999",
      prev_uid: "00000",
      next_uid: "22222",
      route: "route-id",
      rail_line: :red,
      trip_type: "G",
      origin_station: "origin-station",
      destination_station: "destination-station"
    }
  end

  def vehicle_factory do
    %Orbit.Vehicle{
      position: build(:vehicle_position),
      trip_update: build(:trip_update),
      ocs_trips: %{
        current: nil,
        next: []
      }
    }
  end

  def badge_serial_factory do
    %Orbit.BadgeSerial{
      badge_serial: sequence("employee_badge_serial")
    }
  end

  def certification_factory do
    %Orbit.Certification{
      badge: "1234",
      type: :rail,
      rail_line: :blue,
      expires: "2023-12-11"
    }
  end

  def ocs_trip_factory do
    %Orbit.Ocs.Trip{
      service_date: ~D[2025-07-08],
      rail_line: :red,
      uid: "12345678",
      scheduled_departure: ~U[2025-07-08 16:00:00Z],
      scheduled_arrival: ~U[2025-07-08 17:00:00Z],
      origin_station: "ASHMONT",
      destination_station: "ALEWIFE",
      route: "Red",
      offset: 4,
      train_uid: "TRAIN_UID"
    }
  end

  def ocs_train_factory do
    %Orbit.Ocs.Train{
      service_date: ~D[2025-07-08],
      rail_line: :red,
      uid: "5484208E",
      cars: ["1736", "1737", "1724", "1725", "1741", "1740"]
    }
  end

  def vehicle_event_factory do
    %Realtime.Data.VehicleEvent{
      service_date: ~D[2025-07-08],
      rail_line: :red,
      cars: ["1736", "1737", "1724", "1725", "1741", "1740"],
      station_id: "place-asmnl",
      vehicle_id: "5484208E",
      direction_id: 1,
      arrival_departure: :departure,
      timestamp: ~U[2025-07-08 16:05:24Z]
    }
  end
end
